x-pack/plugins/apm/server/routes/fleet/api_keys/add_api_key_to_policies_if_missing.ts

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { omit } from 'lodash';
+import { PackagePolicy } from '@kbn/fleet-plugin/common';
+import { CoreStart, Logger } from '@kbn/core/server';
+import { FleetStartContract } from '@kbn/fleet-plugin/server';
+import { getInternalSavedObjectsClient } from '../../../lib/helpers/get_internal_saved_objects_client';
+import { APMPluginStartDependencies } from '../../../types';
+import { getApmPackagePolicies } from '../get_apm_package_policies';
+import { createApmServerApiKey } from './create_apm_server_api_key';
+import {
+  getPackagePolicyWithApiKey,
+  policyHasApiKey,
+} from '../get_package_policy_decorators';
+
+export async function addApiKeyToPackagePoliciesIfMissing({
+  coreStartPromise,
+  pluginStartPromise,
+  logger,
+}: {
+  coreStartPromise: Promise<CoreStart>;
+  pluginStartPromise: Promise<APMPluginStartDependencies>;
+  logger: Logger;
+}) {
+  const coreStart = await coreStartPromise;
+  const { fleet } = await pluginStartPromise;
+  if (!fleet) {
+    return;
+  }
+
+  const apmFleetPolicies = await getApmPackagePolicies({
+    coreStart,
+    fleetPluginStart: fleet,
+  });
+
+  return Promise.all(
+    apmFleetPolicies.items.map((policy) => {
+      return addApiKeyToPackagePolicyIfMissing({
+        policy,
+        coreStart,
+        fleet,
+        logger,
+      });
+    })
+  );
+}
+
+export async function addApiKeyToPackagePolicyIfMissing({
+  policy,
+  coreStart,
+  fleet,
+  logger,
+}: {
+  policy: PackagePolicy;
+  coreStart: CoreStart;
+  fleet: FleetStartContract;
+  logger: Logger;
+}) {
+  if (policyHasApiKey(policy)) {
+    logger.debug(`Policy (${policy.id}) already has api key`);
+    return;
+  }
+
+  logger.debug(`Policy (${policy.id}) does not have api key`);
+
+  const { api_key: apiKey } = await createApmServerApiKey({
```

Comment:
For api key in apm-server.yml, apm server expects a `id:api_key` value, which is the base64 decoded version of the `encoded` field in the response, or just `id` and `api_key` field concatenated with a `:` separator.

Reply:
Makes sense. I've updated the api key format to include the id and separated by `:`. See the full config for APM Server here:

```json
          "config": {
            "apm-server": {
              "value": {
                "agent_config": [
                  {
                    "service": {},
                    "etag": "0d8423d58d01a04ed7eae94eed94632a8d152314",
                    "config": { "transaction_max_spans": "22" }
                  }
                ],
                "agent": {
                  "config": {
                    "elasticsearch": {
                      "api_key": "k22MhoUBlRqh1NK7iPYK:4ieFNXQeR1mWkBErzgHCCw}"
                    }
                  }
                },
                "rum": {
                  "source_mapping": {
                    "metadata": [
                      {
                        "service.name": "uploading-test",
                        "service.version": "1.0.0",
                        "sourcemap.url": "/api/fleet/artifacts/uploading-test-1.0.0/2f5d4e64ffde4acde832039186ca1652ed315fb0ecbcc1b398677b3bcba521df",
                        "bundle.filepath": "bar"
                      }
                    ],
                    "api_key": "lW2MhoUBlRqh1NK7ifbT:btwmjfVyTMaQKKSJzoc6XQ}"
                  }
                }
              }
            }
          },
```

Reply:
There's a curly brace at the end of `"api_key": "lW2MhoUBlRqh1NK7ifbT:btwmjfVyTMaQKKSJzoc6XQ}"`. Is that a typo? Otherwise it looks good to me

Reply:
Hmm.. Maybe. I think I just copied the full response verbatim. It might be a valid character in the api key.

Reply:
> There's a curly brace at the end of `"api_key": "lW2MhoUBlRqh1NK7ifbT:btwmjfVyTMaQKKSJzoc6XQ}"`. Is that a typo? Otherwise it looks good to me

Great catch! Yes, that was a typo!